File: sdk/ios/Sources/FrakSDK/AppLink/InstallLinks.swift
enum InstallLinks {
    // Storefront-less on purpose: the App Store app resolves it against the user's own storefront,
    // while the same form 404s in a browser outside the territories the app is sold in.
    private static let appStoreURL = "https://apps.apple.com/app/id6759159306"

    /// Links this installation's anonymous id to the user's wallet.
    ///
    /// `installProof` rides as a search param, not a fragment: the wallet's deep-link router
    /// calls `navigate`, so the hash is already gone by the time `/install` renders.
    static func deepLink(
        scheme: String,
        merchantId: String,
        anonymousId: String,
        installProof: String? = nil
    ) -> String {
        let url =
            "\(scheme)://install?m=\(PercentEncoding.encode(merchantId))"
            + "&a=\(PercentEncoding.encode(anonymousId))"
        guard let installProof else { return url }
        return url + "&p=" + PercentEncoding.encode(installProof)
    }

    // No identity carried: iOS has no Play-style install referrer.
    static func appStore() -> String {
        appStoreURL
    }

    /// The same payload as `deepLink`, addressed to the wallet's own universal-link domain
    /// instead of its custom scheme. Tried first: `open(_:options:[.universalLinksOnly: true])`
    /// opens silently on an installed app and answers false rather than falling through to a
    /// Safari tab on one that is not, so failing this rung costs nothing `deepLink` would not
    /// also have to recover from.
    static func universalLink(
        walletOrigin: String,
        merchantId: String,
        anonymousId: String,
        installProof: String? = nil
    ) -> String {
        let url =
            "\(walletOrigin)/install?m=\(PercentEncoding.encode(merchantId))"
            + "&a=\(PercentEncoding.encode(anonymousId))"
        guard let installProof else { return url }
        return url + "#p=" + PercentEncoding.encode(installProof)
    }

    /// The wallet's hosted install page (install code plus store link) — what the sharing sheet
    /// navigates to, as opposed to the store listing `appStore()` returns. The proof rides in the
    /// fragment; `returnScheme`/`sessionId` let the page hand the code back. `clip=host` stops the
    /// page writing the code too: both writes land, and a plain one arriving after this SDK's
    /// `localOnly` and expiring write replaces it.
    static func installPage(
        walletOrigin: String,
        merchantId: String,
        anonymousId: String,
        returnScheme: String,
        sessionId: String,
        proof: String?
    ) -> String {
        let url =
            "\(walletOrigin)/install?embed=native&clip=host"
            + "&m=\(PercentEncoding.encode(merchantId))"
            + "&a=\(PercentEncoding.encode(anonymousId))"
            + "&returnScheme=\(PercentEncoding.encode(returnScheme))"
            + "&sid=\(PercentEncoding.encode(sessionId))"
        guard let proof else { return url }
        return url + "#p=" + PercentEncoding.encode(proof)
    }
}

File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/applink/InstallLinks.kt
// Opted in per file rather than module-wide, so tests still see the marker where a merchant would.
@file:OptIn(InternalFrakApi::class)

package id.frak.sdk.applink

import id.frak.sdk.InternalFrakApi
import id.frak.sdk.net.PercentEncoding
import id.frak.sdk.net.UrlQuery

/** The two URLs that link this installation's anonymous id to a Frak wallet. */
internal object InstallLinks {
    /**
     * `<scheme>://install?m=&a=&p=`. [installProof] rides as a search param, not a fragment: the
     * wallet's deep-link router empties the hash before `/install` renders.
     */
    fun deepLink(
        scheme: String,
        merchantId: String,
        anonymousId: String,
        installProof: String? = null,
    ): String {
        val url =
            "$scheme://install?m=${PercentEncoding.encode(merchantId)}" +
                "&a=${PercentEncoding.encode(anonymousId)}"
        return if (installProof == null) url else "$url&p=${PercentEncoding.encode(installProof)}"
    }

    /**
     * Play Store listing with an install referrer carrying the same pair. The referrer is built as
     * text, not via [UrlQuery], since its own separators must not be re-encoded.
     */
    fun playStore(
        packageId: String,
        merchantId: String,
        anonymousId: String,
        installProof: String? = null,
    ): String {
        val referrer =
            buildString {
                append("merchantId=").append(merchantId)
                append("&anonymousId=").append(anonymousId)
                if (installProof != null) append("&proof=").append(installProof)
            }
        return "$PLAY_STORE_BASE?id=$packageId&referrer=${PercentEncoding.encode(referrer)}"
    }

    /**
     * The wallet's hosted install page: shows the install code and the store link, unlike
     * [playStore] which is the store listing itself. The proof rides in the fragment so it never
     * reaches a server, and `embed=native` is the same marker `/sharing` reads. `clip=host` stops
     * the page writing the code too: both writes land, and a plain one arriving after this SDK's
     * `EXTRA_IS_SENSITIVE` write is the one the system previews.
     */
    fun installPage(
        walletOrigin: String,
        merchantId: String,
        anonymousId: String,
        returnScheme: String,
        sessionId: String,
        proof: String?,
    ): String {
        val url =
            "$walletOrigin/install?embed=native&clip=host" +
                "&m=${PercentEncoding.encode(merchantId)}" +
                "&a=${PercentEncoding.encode(anonymousId)}" +
                "&returnScheme=${PercentEncoding.encode(returnScheme)}" +
                "&sid=${PercentEncoding.encode(sessionId)}"
        return if (proof == null) url else "$url#p=${PercentEncoding.encode(proof)}"
    }

    private const val PLAY_STORE_BASE = "https://play.google.com/store/apps/details"
}
